tryoutmanager/models.py

```diff
@@ -0,0 +1,139 @@
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class Marker(Base):
+    __tablename__ = 'markers'
+
+    email = Column(String, nullable=False, unique=True, primary_key=True)
+    password = Column(String, nullable=False)
+    name = Column(String, nullable=False)
+    challenges = relationship('Challenges',
+                              secondary=markers_challenges,
+                              back_populates='markers')
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class User(Base):
+    __tablename__ = 'users'
+
+    email = Column(String, primary_key=True)
+    password = Column(String, nullable=False)
+    name = Column(String, nullable=False)
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+    confirmed = Column(Boolean, nullable=False, default=False)
+    email_confirmation_slug = Column(String, unique=True)
+    email_confirmation_expiry = Column(DateTime(timezone=True))
+
+
+class Challenge(Base):
+    __tablename__ = 'challenges'
+
+    id = Column(UUID, default=uuid.uuid4, primary_key=True)
+    name = Column(String, nullable=False, unique=True)
+    markers = relationship('Markers',
+                           secondary=markers_challenges,
+                           back_populates='challenges')
+    submissions = relationship('Submission',
+                               secondary=challenges_submissions,
+                               back_populates='submissions')
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class Submission(Base):
+    __tablename__ = 'submissions'
+
+    id = Column(UUID, default=uuid.uuid4, primary_key=True)
+    contestant = relationship('User', back_populates='submissions')
+    contestant_email = Column(String,
+                              ForeignKey('users.email'),
+                              nullable=False)
+    challenge = relationship('Challenge', back_populates='submissions')
+    challenge_id = Column(UUID,
+                          ForeignKey('challenges.id'),
+                          nullable=False)
+    reservation = relationship('Reservation')
+    status = Column(Enum('open',
+                         'closed',
+                         name='submission_status'),
+                    default='closed',
+                    nullable=False)
+
+    # A contestant may only have one submission per challenge
+    __table_args__ = (
+                Index('index_submissions_on_contestant_and_challenge',
+                      'contestant_email',
+                      'challenge_id',
+                      unique=True),
+            )
+
+
+class Reservation(Base):
+    __tablename__ = 'reservations'
+
+    id = Column(UUID, default=uuid.uuid4, primary_key=True)
+    submission = relationship('Submission',
+                              back_populates='reservation',
+                              uselist=False)
+    submission_id = Column(UUID,
```

Comment:
Does a reservation require a submission to be created each time? 
If so we'd need to have the following set of status'es for submissions:
- Booked
- In progress
- Submitted
- Late Submission
- Cancelled

Then, we'd need to create a system for recognising that a contestant didn't complete that submission and overwrite it with a new submission since there is a 1 to 1 association between users and submissions.

I think we could simplify this by creating the submission entry at the start of the window IF it wasn't cancelled. IE we've given them access to the repository. 
This will make more sense down the road to create a submission at the same time as managing the gitolite aspect.

With these changes, then we wouldn't need a complicated setup for the submission status.

Reply:
A reservation creates a submission via a Python-side hook as soon as a reservation's created.

I don't understand how deferring submission creation would necessarily be _simpler_ than being eager about it; could you go into more detail about why?

Reply:
Re: reservation statuses, I'd like to float a three-value enum for `Reservation:` and another three-value enum for `Submission`.

Under `Submission`:
- `in_progress`: default if we use the deferred `Submission`-creation scheme you're suggesting
- `submitted`: updated when a user marks a submission as complete in the UI before the window closes
- `late`: updated when a user marks a submission is _not_ marked as complete by the end of the window.

Under `Reservation`:
- `closed`: default
- `open`: value during the submission window, including late submission times
- `cancelled`: if a reservation is cancelled before the submission window opens.

I'm on the hedge about keeping `cancelled` in; we could just delete reservation records, since AFAIK we don't benefit from keeping those around after cancellation (do we?)

If we take an eager approach to creating `Submission` records, we would add a fourth value to that enum, along the lines of `upcoming`.

Reply:
I agree with the last comment. I spent some time thinking about the user story I had but either way, we'll need to do a bunch of logic to verify the state of whether a submission has been booked completed etc.

Part of me agrees with the philosophy of wanting to delete empty rows but I'd also like to know if people are booking and cancelling frequently to know if there are any issues and reach out to people who might be struggling.